bluesky/tests/test_external_assets_and_paging.py

```diff
@@ -1,479 +1,614 @@
-from typing import Dict, Iterator
-from bluesky import Msg
+from typing import Dict, Iterator, Optional
 from event_model.documents.event_descriptor import DataKey
-from event_model.documents.event_page import PartialEventPage
 from event_model.documents.event import PartialEvent
 from event_model.documents.resource import PartialResource
 from event_model.documents.datum import Datum
 from event_model.documents.stream_datum import StreamDatum
 from event_model.documents.stream_resource import StreamResource
-from bluesky.utils import new_uid, IllegalMessageSequence
-from itertools import chain
+import bluesky.plans as bp
+import bluesky.plan_stubs as bps
 import pytest
 from bluesky.protocols import (
-    Asset, Readable,
-    Reading, SyncOrAsync, WritesExternalAssets, Flyable, EventCollectable, EventPageCollectable, Collectable
+    Asset,
+    Readable,
+    Reading,
+    HasName,
+    WritesExternalAssets,
+    EventCollectable,
+    EventPageCollectable,
+    Collectable,
+    WritesStreamAssets,
+    StreamAsset,
 )
 
 
-def read_Readable(self) -> Dict[str, Reading]:
-    return dict(det2=dict(value=1.2, timestamp=0.0))
+class DocHolder(dict):
```

Comment:
This is clever. 

I'll mention that there is a [utility in the test utils ](https://github.com/bluesky/bluesky/blob/91b4a8a2add41b2cecadc4357b32810437533912/bluesky/tests/utils.py#L31-L50)with a similar functionality, but I like the simplicity of this one.

It _might_ be worth considering using `collections.defaultdict(list)` instead, which I think is a more "mainstream" / familiar way to achieve this, and refactoring `assert_emitted` into a function. On the other hand, I'm not sure that buys much over implementation here. Just a thought.

Reply:
The choice between:
```python
class DocHolder(dict):
    def append(self, name, doc):
        self.setdefault(name, []).append(doc)

    def assert_emitted(self, **numbers: int):
        assert list(self) == list(numbers)
        assert {name: len(d) for name, d in self.items()} == numbers

def test_plan():
    docs = DocHolder()
    RE(plan(), docs.append)
    docs.assert_emitted(start=1, descriptor=1, resource=1, datum=1, event=1, stop=1)
```
and
```python
from collections import defaultdict

def assert_emitted(docs: Dict[str, list], **numbers: int):
    assert list(docs) == list(numbers)
    assert {name: len(d) for name, d in docs.items()} == numbers

def test_plan():
    docs = defaultdict(list)
    RE(plan(), lambda name, doc: docs[name].append(doc))
    assert_emitted(docs, start=1, descriptor=1, resource=1, datum=1, event=1, stop=1)
```

I think I marginally prefer the second, but not enough to change all the tests. @abbiemery I leave the choice up to you...

Reply:
I am in complete agreement. Slight preference, but not sure it's worth the effort to change.

Reply:
Given the time pressure of these now, i'd rather make that a new issue to clean things up if it is preferred.